blockchain/peer_test.go

```diff
@@ -0,0 +1,241 @@
+package blockchain
+
+import (
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	cmn "github.com/tendermint/tendermint/libs/common"
+	"github.com/tendermint/tendermint/libs/log"
+	"github.com/tendermint/tendermint/p2p"
+)
+
+var (
+	numErrFuncCalls int
+	lastErr         error
+	peerTestMtx     sync.Mutex
+)
+
+func resetErrors() {
+	peerTestMtx.Lock()
+	defer peerTestMtx.Unlock()
+	numErrFuncCalls = 0
+	lastErr = nil
+}
+
+func errFunc(err error, peerID p2p.ID) {
+	peerTestMtx.Lock()
+	defer peerTestMtx.Unlock()
+	_ = peerID
+	lastErr = err
+	numErrFuncCalls++
+}
+
+// check if peer timer is running or not (a running timer can be successfully stopped)
+// Note: it does stop the timer!
+func checkByStoppingPeerTimer(t *testing.T, peer *bpPeer, running bool) {
+	assert.NotPanics(t, func() {
+		stopped := peer.timeout.Stop()
+		if running {
+			assert.True(t, stopped)
+		} else {
+			assert.False(t, stopped)
+		}
+	})
+}
+
+func TestPeerResetMonitor(t *testing.T) {
+	peer := &bpPeer{
+		id:      p2p.ID(cmn.RandStr(12)),
+		height:  10,
+		logger:  log.TestingLogger(),
+		errFunc: errFunc,
+	}
+	peer.resetMonitor()
+	assert.NotNil(t, peer.recvMonitor)
+}
+
+func TestPeerTimer(t *testing.T) {
+	peerTimeout = 2 * time.Millisecond
+
+	peer := &bpPeer{
+		id:      p2p.ID(cmn.RandStr(12)),
+		height:  10,
+		logger:  log.TestingLogger(),
+		errFunc: errFunc,
+	}
+	assert.Nil(t, peer.timeout)
+
+	// initial reset call with peer having a nil timer
+	peer.resetTimeout()
+	assert.NotNil(t, peer.timeout)
+	// make sure timer is running and stop it
+	checkByStoppingPeerTimer(t, peer, true)
+
+	// reset with non nil expired timer
+	peer.resetTimeout()
+	assert.NotNil(t, peer.timeout)
+	// make sure timer is running and stop it
+	checkByStoppingPeerTimer(t, peer, true)
+	resetErrors()
+
+	// reset with running timer (started above)
+	time.Sleep(time.Millisecond)
+	peer.resetTimeout()
+	assert.NotNil(t, peer.timeout)
+
+	// let the timer expire and ...
+	time.Sleep(3 * time.Millisecond)
+	checkByStoppingPeerTimer(t, peer, false)
+
+	peerTestMtx.Lock()
+	// ... check an error has been sent, error is peerNonResponsive
+	assert.Equal(t, 1, numErrFuncCalls)
+	assert.Equal(t, lastErr, errNoPeerResponse)
+	peerTestMtx.Unlock()
+
+	// Restore the peerTimeout to its original value
+	peerTimeout = 15 * time.Second
```

Comment:
Do we really need to restore this value? Every test should set it initially.

Reply:
Ok I added the const for now. This was a pain for me as I would forget this set at msec and other tests (reactor_test.go) would fail when running on the full package. I think it should be included as a field in reactor or pool and in general I would love to get rid of all global vars.

Reply:
removed.